tools/internal/parser/parser.go

```diff
@@ -0,0 +1,559 @@
+// Package parser implements a validating parser for the PSL files.
+package parser
+
+import (
+	"net/mail"
+	"net/url"
+	"strings"
+)
+
+// Parse parses src as a PSL file and returns the parse result.
+//
+// The parser tries to keep going when it encounters errors. Parse and
+// validation errors are accumulated in the Errors field of the
+// returned File.
+//
+// If the returned File has a non-empty Errors field, the parsed file
+// may not be valid, and should not be used to make decisions about
```

Comment:
Is this actually a may? Don't parse errors always mean that it is invalid?

Reply:
I guess it depends on your definition: errors can be hard errors that cause incorrect suffix calculations, or they can be lint errors that only matter if you're trying to submit a PR.

... But thinking about it more, for now this parser is 100% for the second purpose, so all errors are "bad". I've rewritten the paragraph to try and be more precise.